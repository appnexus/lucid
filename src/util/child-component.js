import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {

			/**
			 * getOwnChildren
			 *
			 * Returns all collective children of elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} All the collective children of elements with this component type
			 */
			getOwnChildren(elements) {
				let component = this;
				let children = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						children = children.concat(React.Children.map(element.props.children, x => x));
					}
				});
				return children;
			},

			/**
			 * getAll
			 *
			 * Returns all elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} elements with this component type
			 */
			getAll(elements) {
				let component = this;
				let elementsOfThisType = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						elementsOfThisType = elementsOfThisType.concat(element);
					}
				});
				return elementsOfThisType;
			},

			/**
			 * getAllProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} `props` objects of elements with this component type
			 */
			getAllProps(elements) {
				let component = this;
				return _.map(component.getAll(elements), 'props');
			},

			getAllAsProps(props, elements) {
				let component = this;
				return _.map(_.flatten([_.get(props, definition.propName, [])]), (childProp) => {
					if (_.isPlainObject(childProp)) {
						return _.assign({}, childProp, {children: childProp.children || childProp.text});
					} else {
						return {
							children: childProp
						};
					}
				}).concat(component.getAllProps(elements));
			}
		}
	});
}
