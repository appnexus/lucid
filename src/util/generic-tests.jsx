import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
