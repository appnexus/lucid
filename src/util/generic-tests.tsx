import sinon from 'sinon';
import { parse } from 'path';
import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow, render } from 'enzyme';
import assert from 'assert';
import _, { each, omit, keys, includes } from 'lodash';
import glob from 'glob';
import { addons, mockChannel } from '@storybook/addons';
import timekeeper from 'timekeeper';

addons.setChannel(mockChannel());

interface ICommonConfig {
	getDefaultProps?: () => {};
	exemptFunctionProps?: string[];
	exemptChildComponents?: string[];
	selectRoot?: any;
	noExport?: boolean;
}

// Common tests for all our components
export function common(Component: any, config: ICommonConfig = {} as any) {
	const {
		getDefaultProps = _.constant({}),
		exemptFunctionProps = [],
		exemptChildComponents = [],
		selectRoot = _.identity,
		noExport = false,
	} = config;

	function generateDefaultProps(props = {}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		if (!Component) {
			throw new Error('An undefined component was passed to generic tests.');
		}

		if (Component._isLucidHybridComponent) {
			throw new Error(
				`You're trying to run generic tests on a hybrid component which is bad and won't work and will make you cry. Check your spec files for ${Component.displayName} and import the raw component instead of the hybrid version.`
			);
		}

		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			const style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(
				<Component {...generateDefaultProps()} style={style} />,
				{ disableLifecycleMethods: true }
			);

			const rootWrapper = selectRoot(wrapper).first();
			const rootStyle = rootWrapper.prop('style');
			assert(
				_.every(style, (val, key) => val === rootStyle[key]),
				'root style must contain passed styles'
			);
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(
				<Component {...generateDefaultProps()} className={expectedClass} />,
				{ disableLifecycleMethods: true }
			);
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${expectedClass}'`
			);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${Component.displayName}'`
			);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '));
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, (className) => {
				assert(
					_.includes(className, `lucid-${Component.displayName}`),
					`${className} must be scoped`
				);
			});
		});

		describe('function propTypes', () => {
			const funcProps = _.pickBy(
				Component.propTypes,
				(propType) => propType === PropTypes.func
			);

			_.forEach(funcProps, (propType, propName) => {
				it(`${propName} should only use onX convention for function proptypes`, () => {
					assert(
						_.startsWith(propName, 'on') ||
							_.includes(exemptFunctionProps, propName),
						`${propName} must follow onX convention`
					);
				});
			});
		});

		describe('child components', () => {
			// Child components are all function types which start with a capital letter
			const childComponents = _.pickBy(Component, (value, key) => {
				return /^[A-Z]/.test(key) && _.isFunction(value);
			});

			describe('propNames in propTypes', () => {
				_.flow(
					(x) => _.map(x, 'propName'),
					(x) => _.compact(x),
					(x) => _.flatMap(x, _.castArray),
					(x) =>
						_.reject(x, (propName) =>
							_.includes(exemptChildComponents, propName)
						),
					(x) =>
						_.forEach(x, (propName) => {
							it(`should include ${propName} in propTypes`, () => {
								assert(
									Component.propTypes[propName],
									`must include ${propName} in propTypes`
								);
							});
						})
				)(childComponents);
			});
		});

		describe('example testing', () => {
			const fileNames = glob.sync(
				`./src/components/**/${Component.displayName}/*.stories.@(j|t)sx`
			);
			each(fileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				each(omit(lib, ['default']), (Story, name) => {
					it(`should match snapshot(s) for ${name}`, () => {
						let result;
						try {
							result = render(<Story {...Story.args} />, {
								disableLifecycleMethods: true,
							});
							expect(result).toMatchSnapshot();
						} catch (err) {
							expect(err).toMatchSnapshot();
						}
					});
				});
			});

			// Support for older examples
			const exampleFileNames = glob.sync(
				`./src/components/**/${Component.displayName}/examples/*.@(j|t)sx`
			);
			_.each(exampleFileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				const Example = lib.default;

				const title = parse(path).name;
				it(`should match snapshot(s) for ${title}`, () => {
					const shallowExample = shallow(<Example />, {
						disableLifecycleMethods: true,
					});

					// If the root of the example is an instance of the Component under test, snapshot it.
					// Otherwise, look under the root for instances of the Component and snapshot those.
					if (shallowExample.is(Component.displayName)) {
						expect(
							shallow(<Component {...shallowExample.props()} />, {
								disableLifecycleMethods: true,
							})
						).toMatchSnapshot();
					} else {
						shallowExample.find(Component.displayName).forEach((example) => {
							expect(
								shallow(<Component {...example.props()} />, {
									disableLifecycleMethods: true,
								})
							).toMatchSnapshot();
						});
					}
				});
			});
		});
	});
}

// Common tests for all our icon components

interface IIconConfig {
	includeInitialState: boolean;
}

export function icons(Component: any, config: IIconConfig = {} as any) {
	// The default expectation is for every Icon to omit `initailState`,
	// if it is passed through to the underlying element
	const { includeInitialState = false } = config;

	describe('[icon]', () => {
		it('should almost always omit the `initialState` key', () => {
			const wrapper = shallow(<Component initialState={{ testState: true }} />);
			const rootProps = keys(wrapper.first().props());

			expect(includes(rootProps, 'initialState')).toBe(includeInitialState);
		});

		it('should add the correct class for isClickable', () => {
			const wrapper = mount(<Component isClickable={true} />);
			const targetClassName = 'lucid-Icon-is-clickable';
			assert(
				wrapper.find('svg').hasClass(targetClassName),
				`Missing '${targetClassName}' class`
			);
		});
	});
}

// Common tests for all control components
export function controls(
	Component: any,
	{ callbackName, controlSelector, eventType, additionalProps = {} }: any
) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialprop: expectedSpecialProp,
				[callbackName]: sinon.spy(),
				...additionalProps,
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).first().simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const {
				props: { specialprop },
				event,
			}: any = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(
				specialprop,
				expectedSpecialProp,
				'incorrect or missing specialProp'
			);
		});
	});
}

// Common tests for all Functional Components
//
// These tests are intended to help us make sure our FCs are shaped corrected.
// They are necessary because there isn't a perfect way to get the defaultProps
// to be factored in correctly yet with React/TypeScript:
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30695#issuecomment-474780159
export function functionalComponents(FC: any) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[functionalComponent]', () => {
		it('should have the correct `peek` properties', () => {
			expect(FC.propName === undefined || typeof FC.propName === 'string').toBe(
				true
			);
			expect(
				FC._isPrivate === undefined || typeof FC._isPrivate === 'boolean'
			).toBe(true);

			expect(typeof FC.peek).toBe('object');
			expect(typeof FC.peek.description).toBe('string');
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.categories === undefined || Array.isArray(FC.peek.categories)
			).toBe(true);
			expect(
				FC.peek.madeFrom === undefined || Array.isArray(FC.peek.madeFrom)
			).toBe(true);
		});
	});
}

export const mockDate = (dateString: string) => {
	timekeeper.freeze(new Date(dateString));
};
