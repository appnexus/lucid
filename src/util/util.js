import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
