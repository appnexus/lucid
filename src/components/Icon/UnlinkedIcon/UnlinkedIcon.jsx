import _ from 'lodash';
import React from 'react';
import Icon from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';
import { createClass as createReactClass } from '../../../util/component-types';

const cx = lucidClassNames.bind('&-UnlinkedIcon');

/**
 *
 * {"categories": ["visual design", "icons"], "extend": "Icon", "madeFrom": ["Icon"]}
 *
 * For all those times you just need to break away.
 */
const UnlinkedIcon = createReactClass({
	displayName: 'UnlinkedIcon',
	propTypes: {
		...Icon.propTypes,
	},

	render() {
		const { className, ...passThroughs } = this.props;

		return (
			<Icon
				{...passThroughs}
				{..._.pick(passThroughs, _.keys(Icon.propTypes))}
				className={cx('&', className)}
			>
				<path d="M11.826,1.791c0.564,0,1.383,0.215,1.779,0.613c0.393,0.391,0.617,1.003,0.617,1.679c0,0.675-0.225,1.288-0.615,1.681 l-3.406,3.209l1.986,0.004l2.41-2.222c1.473-1.474,1.473-3.871,0-5.345c-0.711-0.712-1.66-1.105-2.672-1.105 S9.967,0.697,9.258,1.411L7.032,3.557L7.023,5.523l3.227-3.158C10.609,2.005,11.197,1.791,11.826,1.791z" />
				<path d="M5.373,13.598c-0.398,0.396-0.924,0.613-1.48,0.613c-0.558,0-1.083-0.217-1.479-0.611 c-0.815-0.816-0.816-2.545-0.002-3.359l3.101-3.234L3.551,7L1.42,9.248c-0.714,0.711-1.106,1.658-1.106,2.67 s0.393,1.961,1.104,2.672c0.711,0.713,1.66,1.105,2.673,1.105c1.011,0,1.959-0.393,2.671-1.105l2.244-2.146L9.002,10.49 L5.373,13.598z" />
				<path d="M15.98,10.127c-0.047-0.055-0.117-0.086-0.188-0.082l-3.631,0.045c-0.121,0.002-0.223,0.092-0.236,0.213 c-0.016,0.119,0.057,0.232,0.174,0.266l3.525,1.051c0.023,0.006,0.045,0.008,0.066,0.008c0.049,0,0.094-0.012,0.135-0.039 c0.059-0.039,0.1-0.102,0.107-0.17l0.074-1.1C16.02,10.25,16.027,10.18,15.98,10.127z" />
				<path d="M10.24,11.58c-0.117,0.016-0.209,0.115-0.211,0.234l-0.021,3.631c0,0.072,0.029,0.141,0.082,0.189 c0.045,0.039,0.102,0.061,0.162,0.061c0.01,0,0.021,0,0.031-0.002l1.072-0.104c0.068-0.01,0.131-0.049,0.17-0.109 c0.039-0.059,0.051-0.133,0.031-0.201l-1.051-3.523C10.475,11.637,10.361,11.562,10.24,11.58z" />
				<path d="M5.384,4.132c0.029,0.105,0.127,0.176,0.232,0.176c0.012,0,0.023-0.001,0.033-0.002c0.12-0.016,0.209-0.117,0.212-0.237 l0.045-3.631c0.001-0.071-0.028-0.14-0.082-0.187C5.773,0.204,5.7,0.183,5.631,0.192L4.536,0.297 C4.465,0.307,4.403,0.347,4.364,0.407c-0.039,0.059-0.1,0.359-0.08,0.428L5.384,4.132z" />
				<path d="M4.092,5.744c0.015-0.12-0.059-0.233-0.175-0.267L0.391,4.429c-0.067-0.021-0.142-0.01-0.2,0.029 C0.131,4.497,0.09,4.559,0.082,4.63L0.008,5.725c-0.011,0.071-0.019,0.142,0.027,0.195C0.083,5.972,0.149,6,0.219,6h0.004 l3.631-0.046C3.975,5.954,4.075,5.863,4.092,5.744z" />
			</Icon>
		);
	},
});

export default UnlinkedIcon;
