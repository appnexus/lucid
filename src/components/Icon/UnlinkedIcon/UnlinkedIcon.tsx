import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import Icon, { IIconProps } from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';

const cx = lucidClassNames.bind('&-UnlinkedIcon');

export const iconPropTypes = {
	/** Size variations of the icons. \`size\` directly effects height and width
		but the developer should also be conscious of the relationship with
		\`viewBox\`. */
	size: PropTypes.number,

	/** Size handles width and height, whereas \`width\` can manually override the width that would be set by size. */
	width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

	/** Size handles width and height, whereas \`height\` can manually override the height that would be set by size. */
	height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

	/** \`viewBox\` is very important for SVGs. You can think of \`viewBox\` as
		the "artboard" for our SVG while \`size\` is the presented height and
		width. */
	viewBox: PropTypes.string,

	/** Sets the color of the Icon.  May not be applicable for icons that are tied
		to specific colors (e.g. DangerIcon). */
	color: PropTypes.oneOf([
		'neutral-dark',
		'neutral-light',
		'neutral-extra-light',
		'primary',
		'white',
		'success',
		'warning',
		'secondary-one',
		'secondary-two',
		'secondary-three',
		'secondary-five',
	]),

	/** Any valid SVG aspect ratio. */
	aspectRatio: PropTypes.string,

	/** Adds styling that makes the icon appear clickable. */
	isClickable: PropTypes.bool,

	/** Adds styling that makes the icon appear disabled.  Also forces
		isClickable to be false. */
	isDisabled: PropTypes.bool,

	/** Called when the user clicks the \`Icon\`. Signature:
		\`({event, props}) => {}\` */
	onClick: PropTypes.func,

	/** Called when the user clicks an active, clickable \`Icon\`. Signature:
		\`({event, props}) => {}\` */
	onSelect: PropTypes.func,

	/** Any valid React children. */
	children: PropTypes.element,

	/** Classes that are appended to the component defaults. This prop is run
		through the \`classnames\` library. */
	className: PropTypes.string,
};

export const UnlinkedIcon = ({ className, ...passThroughs }: IIconProps) => {
	return (
		<Icon
			{..._.omit(passThroughs, ['initialState'])}
			className={cx('&', className)}
		>
			<path d='M5.5 3.5l-1-2m8 9l2 1m-4 1l1 2m-8-9l-2-1' />
			<path d='M10.232 7.768a2.993 2.993 0 0 0 2.979-.737l1.414-1.414a3.009 3.009 0 0 0 0-4.243 3.009 3.009 0 0 0-4.243 0L8.968 2.789c-.804.804-1.04 1.956-.737 2.979M5.769 8.231c-1.023-.303-2.176-.067-2.98.737l-1.414 1.414a3.009 3.009 0 0 0 0 4.243 3.009 3.009 0 0 0 4.243 0l1.414-1.414c.804-.804 1.04-1.956.737-2.979' />
		</Icon>
	);
};

UnlinkedIcon.displayName = 'UnlinkedIcon';

UnlinkedIcon.propTypes = iconPropTypes;

UnlinkedIcon.defaultProps = Icon.defaultProps;

export default UnlinkedIcon;
