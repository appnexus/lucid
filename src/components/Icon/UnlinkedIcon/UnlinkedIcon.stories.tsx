import React from 'react';
import { Story, Meta } from '@storybook/react';
import { IIconProps } from '../Icon';
import { UnlinkedIcon } from './UnlinkedIcon';

export default {
	title: 'Icons/Icons/UnlinkedIcon',
	component: UnlinkedIcon,
} as Meta;

//👇 We create a “template” of how args map to rendering
const Template: Story<IIconProps> = (args) => <UnlinkedIcon {...args} />;

//👇 Each story then reuses that template
export const Primary = Template.bind({});
