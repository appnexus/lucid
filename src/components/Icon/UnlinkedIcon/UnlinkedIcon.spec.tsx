import { icons, common } from '../../../util/generic-tests';

import UnlinkedIcon from './UnlinkedIcon';

describe('UnlinkedIcon', () => {
	common(UnlinkedIcon);
	icons(UnlinkedIcon);
});
