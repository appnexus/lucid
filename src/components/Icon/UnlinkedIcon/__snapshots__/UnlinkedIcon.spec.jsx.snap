// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnlinkedIcon [common] example testing should match snapshot(s) for 1.default 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-UnlinkedIcon"
  isBadge={false}
  isClickable={false}
  isDisabled={false}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M11.826 1.79c.564 0 1.383.216 1.78.614.392.39.616 1.003.616 1.68 0 .674-.225 1.287-.615 1.68L10.2 8.974l1.987.003 2.41-2.222c1.473-1.474 1.473-3.87 0-5.345-.71-.712-1.66-1.105-2.672-1.105S9.967.697 9.258 1.41L7.032 3.558l-.01 1.966 3.228-3.158c.36-.36.947-.574 1.576-.574zM5.373 13.598c-.398.396-.924.613-1.48.613-.558 0-1.083-.216-1.48-.61-.814-.816-.815-2.545 0-3.36l3.1-3.233L3.55 7 1.42 9.248c-.714.71-1.106 1.658-1.106 2.67s.393 1.96 1.104 2.672c.71.713 1.66 1.105 2.673 1.105 1.012 0 1.96-.393 2.672-1.105l2.244-2.146-.004-1.954-3.63 3.108zM15.98 10.127c-.047-.055-.117-.086-.188-.082l-3.63.045c-.122.002-.224.092-.237.213-.016.12.057.232.174.266l3.524 1.05c.023.006.045.008.066.008.05 0 .094-.012.135-.04.06-.038.1-.1.107-.17l.074-1.1c.014-.068.02-.138-.026-.19zM10.24 11.58c-.117.016-.21.115-.21.234l-.022 3.63c0 .073.03.142.082.19.045.04.102.06.162.06h.03l1.073-.105c.068-.01.13-.05.17-.11.04-.06.05-.133.03-.2l-1.05-3.524c-.03-.12-.144-.194-.265-.176zM5.384 4.132c.03.105.127.176.232.176.012 0 .023 0 .033-.002.12-.016.208-.117.21-.237L5.907.437c0-.07-.028-.14-.082-.187C5.774.205 5.7.184 5.63.193L4.537.297c-.07.01-.133.05-.172.11-.04.06-.1.36-.08.428l1.1 3.297zM4.092 5.744c.015-.12-.06-.233-.175-.267L.39 4.43c-.066-.022-.14-.01-.2.028-.06.04-.1.1-.108.172L.008 5.725c-.01.07-.02.142.027.195.048.052.114.08.184.08h.003l3.63-.046c.122 0 .222-.09.24-.21z"
  />
</Icon>
`;

exports[`UnlinkedIcon [common] example testing should match snapshot(s) for 2.clickable 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-UnlinkedIcon"
  isBadge={false}
  isClickable={true}
  isDisabled={false}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M11.826 1.79c.564 0 1.383.216 1.78.614.392.39.616 1.003.616 1.68 0 .674-.225 1.287-.615 1.68L10.2 8.974l1.987.003 2.41-2.222c1.473-1.474 1.473-3.87 0-5.345-.71-.712-1.66-1.105-2.672-1.105S9.967.697 9.258 1.41L7.032 3.558l-.01 1.966 3.228-3.158c.36-.36.947-.574 1.576-.574zM5.373 13.598c-.398.396-.924.613-1.48.613-.558 0-1.083-.216-1.48-.61-.814-.816-.815-2.545 0-3.36l3.1-3.233L3.55 7 1.42 9.248c-.714.71-1.106 1.658-1.106 2.67s.393 1.96 1.104 2.672c.71.713 1.66 1.105 2.673 1.105 1.012 0 1.96-.393 2.672-1.105l2.244-2.146-.004-1.954-3.63 3.108zM15.98 10.127c-.047-.055-.117-.086-.188-.082l-3.63.045c-.122.002-.224.092-.237.213-.016.12.057.232.174.266l3.524 1.05c.023.006.045.008.066.008.05 0 .094-.012.135-.04.06-.038.1-.1.107-.17l.074-1.1c.014-.068.02-.138-.026-.19zM10.24 11.58c-.117.016-.21.115-.21.234l-.022 3.63c0 .073.03.142.082.19.045.04.102.06.162.06h.03l1.073-.105c.068-.01.13-.05.17-.11.04-.06.05-.133.03-.2l-1.05-3.524c-.03-.12-.144-.194-.265-.176zM5.384 4.132c.03.105.127.176.232.176.012 0 .023 0 .033-.002.12-.016.208-.117.21-.237L5.907.437c0-.07-.028-.14-.082-.187C5.774.205 5.7.184 5.63.193L4.537.297c-.07.01-.133.05-.172.11-.04.06-.1.36-.08.428l1.1 3.297zM4.092 5.744c.015-.12-.06-.233-.175-.267L.39 4.43c-.066-.022-.14-.01-.2.028-.06.04-.1.1-.108.172L.008 5.725c-.01.07-.02.142.027.195.048.052.114.08.184.08h.003l3.63-.046c.122 0 .222-.09.24-.21z"
  />
</Icon>
`;

exports[`UnlinkedIcon [common] example testing should match snapshot(s) for 3.disabled 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-UnlinkedIcon"
  isBadge={false}
  isClickable={false}
  isDisabled={true}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M11.826 1.79c.564 0 1.383.216 1.78.614.392.39.616 1.003.616 1.68 0 .674-.225 1.287-.615 1.68L10.2 8.974l1.987.003 2.41-2.222c1.473-1.474 1.473-3.87 0-5.345-.71-.712-1.66-1.105-2.672-1.105S9.967.697 9.258 1.41L7.032 3.558l-.01 1.966 3.228-3.158c.36-.36.947-.574 1.576-.574zM5.373 13.598c-.398.396-.924.613-1.48.613-.558 0-1.083-.216-1.48-.61-.814-.816-.815-2.545 0-3.36l3.1-3.233L3.55 7 1.42 9.248c-.714.71-1.106 1.658-1.106 2.67s.393 1.96 1.104 2.672c.71.713 1.66 1.105 2.673 1.105 1.012 0 1.96-.393 2.672-1.105l2.244-2.146-.004-1.954-3.63 3.108zM15.98 10.127c-.047-.055-.117-.086-.188-.082l-3.63.045c-.122.002-.224.092-.237.213-.016.12.057.232.174.266l3.524 1.05c.023.006.045.008.066.008.05 0 .094-.012.135-.04.06-.038.1-.1.107-.17l.074-1.1c.014-.068.02-.138-.026-.19zM10.24 11.58c-.117.016-.21.115-.21.234l-.022 3.63c0 .073.03.142.082.19.045.04.102.06.162.06h.03l1.073-.105c.068-.01.13-.05.17-.11.04-.06.05-.133.03-.2l-1.05-3.524c-.03-.12-.144-.194-.265-.176zM5.384 4.132c.03.105.127.176.232.176.012 0 .023 0 .033-.002.12-.016.208-.117.21-.237L5.907.437c0-.07-.028-.14-.082-.187C5.774.205 5.7.184 5.63.193L4.537.297c-.07.01-.133.05-.172.11-.04.06-.1.36-.08.428l1.1 3.297zM4.092 5.744c.015-.12-.06-.233-.175-.267L.39 4.43c-.066-.022-.14-.01-.2.028-.06.04-.1.1-.108.172L.008 5.725c-.01.07-.02.142.027.195.048.052.114.08.184.08h.003l3.63-.046c.122 0 .222-.09.24-.21z"
  />
</Icon>
`;
